File: fe/fe-core/src/main/java/com/starrocks/common/util/DnsCache.java
// Copyright 2021-present StarRocks, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.starrocks.common.util;

import org.apache.commons.validator.routines.InetAddressValidator;

import java.net.InetAddress;
import java.net.UnknownHostException;

// This class is used to translate hostname to IP.
// Because JVM has DNS cache already, this class doesn't maintain its own cache.
public class DnsCache {
    public static String lookup(String hostname) throws UnknownHostException {
        if (InetAddressValidator.getInstance().isValidInet4Address(hostname)) {
            return hostname;
        }
        return InetAddress.getByName(hostname).getHostAddress();
    }

    // This function will try to look up the given hostname. If there is an exception, this function will
    // return the input hostname directly.
    public static String tryLookup(String hostname) {
        try {
            return lookup(hostname);
        } catch (UnknownHostException e) {
            return hostname;
        }
    }
}

File: be/src/util/dns_cache.cpp
// Copyright 2021-present StarRocks, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "util/dns_cache.h"

#include <mutex>
#include <unordered_set>

#include "service/backend_options.h"
#include "util/network_util.h"

namespace starrocks {

DNSCache::DNSCache() {}

DNSCache::~DNSCache() {}

StatusOr<std::string> DNSCache::get(const std::string& hostname) {
    {
        std::shared_lock<std::shared_mutex> lock(mutex);
        auto it = cache.find(hostname);
        if (it != cache.end()) {
            return it->second;
        }
    }
    // Update if not found
    RETURN_IF_ERROR(update(hostname));
    {
        std::shared_lock<std::shared_mutex> lock(mutex);
        return cache[hostname];
    }
}

Status DNSCache::update(const std::string& hostname) {
    std::string ip = "";
    RETURN_IF_ERROR(hostname_to_ip(hostname, ip, BackendOptions::is_bind_ipv6()));
    std::unique_lock<std::shared_mutex> lock(mutex);
    auto it = cache.find(hostname);
    if (it == cache.end() || it->second != ip) {
        cache[hostname] = ip;
        LOG(INFO) << "update hostname " << hostname << "'s ip to " << ip;
    }
    return Status::OK();
}

void DNSCache::refresh() {
    std::unordered_set<std::string> keys;
    {
        std::shared_lock<std::shared_mutex> lock(mutex);
        std::transform(cache.begin(), cache.end(), std::inserter(keys, keys.end()),
                       [](const auto& pair) { return pair.first; });
    }
    for (auto& key : keys) {
        auto st = update(key);
        if (!st.ok()) {
            LOG(WARNING) << "update hostname " << key << " failed when refreshing dns cache";
        }
    }
}

} // namespace starrocks